### PR TITLE
apimachinery: add StringArrayOrString with json/yaml/proto marshaling

### PR DESCRIPTION
**Pull request by sttts**
In analogy to `IntOrString` this allows to have either a string or a string array in JSON/YAML/proto, needed by CRD validation https://github.com/kubernetes/kubernetes/pull/47263.

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/util/arraystr/arraystr.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	}
+	return fmt.Sprintf("%v", arraystr.ArrayVal)
+}
+
+// ArrayValue returns the array value if type Array, or if
+// it is a String, will return a one-element slice.
+func (arraystr *StringArrayOrString) ArrayValue() []string {
+	if arraystr.Type == String {
+		if arraystr.StrVal == "" {
+			return nil
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/util/arraystr/arraystr.go]: Why nil instead of a zero length slice?

File: staging/src/k8s.io/apimachinery/pkg/util/arraystr/arraystr_test.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+	SAOrSOmitEmpty *StringArrayOrString `json:"val2,omitempty"`
+}
+
+func TestStringArrayOrStringUnmarshalJSON(t *testing.T) {
+	zero := FromString("")
+	cases := []struct {
+		input  string
+		result StringArrayOrStringHolder
+	}{
+		{`{"val1": []}`, StringArrayOrStringHolder{SAOrS: FromString("")}},
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/util/arraystr/arraystr_test.go]: This surprised me.  Serialized as an array, not a string?

Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/util/arraystr/arraystr_test.go]: > This surprised me. Serialized as an array, not a string?

Oh, unmarshal test

Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/util/arraystr/arraystr_test.go]: Add a test unmarshalling an empty string please.

**Comment by deads2k** [comment]
Request for an additional test.  lgtm otherwise

/approve no-issue

**Comment by sttts** [comment]
Test added.

**Comment by deads2k** [comment]
/lgtm

@smarterclayton you should look.

**Comment by smarterclayton** [comment]
Given all the pain and suffering on the open api side recently about how painful IntToString is and not easy to represent - does this take that into account?

**Comment by sttts** [comment]
> how painful IntToString is and not easy to represent - does this take that into account?

What's the alternative? To my understanding the openapi trouble is mainly because we create wrong openapi specs, not taking `null` values into account.

About the type, we basically have no choice. The choice was taken by the JSON Schema developers. We only 1:1 embed an OpenAPI spec in our types.

**Comment by sttts** [comment]
@smarterclayton what do you suggest?

**Comment by enisoc** [comment]
Another alternative is to say our "JSON Schema" differs slightly from the official spec, such as having to always use a string array in fields such as this, even if you only have one string (assuming there is no semantic difference between a string and an array with one entry). OpenAPI does this too; their docs explain the minor differences between OpenAPI Schema and the JSON Schema spec.

The ability to exert such control over our own API was after all one of the justifications for importing the structs. If we wanted to match the spec 100%, it may have made more sense to just store the JSON Schema as a string.

**Comment by sttts** [comment]
@enisoc I am still not convinced that something like StringArrayOrString is bad. What are the issues that everybody expects?

Short term in fact we could just fall back to "string only". This is what OpenAPI v3 does and we only support v3 now anyway. But IMO this is a very stupid restriction in the spec and I expect very much hat v3+ will support more than one type. But by that time we can reconsider.

**Comment by enisoc** [comment]
@sttts I was mostly just trying to give a hypothetical answer to the question, "What's the alternative?" I'm not familiar enough with the pain experienced with `IntOrString` to have an opinion on whether `StringArrayOrString` will cause more pain.

@smarterclayton You mentioned pain with OpenAPI. Were you referring to [this thread](https://groups.google.com/d/msg/kubernetes-sig-api-machinery/QcZG6UIrEgk/5Dof3PC9BwAJ)? It seems like that issue will go away eventually when we switch to OpenAPI v3.

**Comment by sttts** [comment]
> It seems like that issue will go away eventually when we switch to OpenAPI v3.

We still don't have the type field as array in OpenAPI v3. We have dropped this now from our JSON Schema types. So the unmarshaler will complain if you try an array. If v4 finally adds arrays, we have to reconsider.

**Comment by enisoc** [comment]
@sttts In the thread, @mbohlool suggested you could work around the specific issue raised there with `oneOf`. Would that not work?

**Comment by sttts** [comment]
> @mbohlool suggested you could work around the specific issue raised there with oneOf. Would that not work?

We are on the other side of the problem: we offer the API, but our users won't be able to use an array for types. Actually, it's not `oneOf`, but `anyOf` that can replace the array, it's just ugly. Here we did a comparison of all those variants of JSON Schema: https://docs.google.com/spreadsheets/d/1Mkm9L7CXGvRorV0Cr4Vwfu0DH7XRi24YHPiDK1NZWo4/edit?usp=sharing

**Comment by enisoc** [comment]
Sorry if I'm creating confusion. I didn't mean to suggest that oneOf/anyOf could fix the problem faced by CRD validation.

I was trying to say, if it's true that (1) the problem in that mailing list thread is the only known pain caused by IntOrString, and (2) the problem in that thread could be resolved by going to OpenAPI V3, then it would imply that there are actually no known, long-term problems with IntOrString. Therefore we would have no reason to suspect adding StringArrayOrString will cause long-term pain.

**Comment by mml** [comment]
cc @mbohlool @jpbetz

**Comment by bgrant0607** [comment]
@sttts We have absolutely no obligation to allow all of JSON / JSONSchema. The K8s API uses its own conventions, and even some of those we regret. Has anyone attempted to determine which features are exercised by the current K8s API surface?

Are there any examples of StringArrayOrString in the K8s core v1 API? IIRC, we explicitly decided not to support that for command, for instance.

We do have the problem that not all of the K8s API can be described by the current OpenAPI spec. At least that was true when I last checked.

cc @mbohlool

**Comment by sttts** [comment]
> We have absolutely no obligation to allow all of JSON / JSONSchema. The K8s API uses its own conventions

That's a good second view point. We use OpenAPI v3 to decribe kube's API (lower bound), but here we want to define the JSON Schema subset to specify APIs following the kube API conventions. Wondering whether we can derive any upper bound from the conventions we envisions also for user-provided CRDs. Sounds like a nice topic for bachelor thesis (hint!). kubeapi-lint.

**Comment by sttts** [comment]
> no known, long-term problems with IntOrString

At least not from the spec point of view: `{anyOf: [{type: "string"}, {type: "integer"}]}` should be equivalent to `{type:["string", "integer"]}`.

**Comment by mbohlool** [comment]
> kubeapi-lint.

https://github.com/kubernetes/kubernetes/issues/44005

**Comment by mbohlool** [comment]
> At least not from the spec point of view: {anyOf: [{type: "string"}, {type: "integer"}]} should be equivalent to {type:["string", "integer"]} in v3, not possible in v2.

`{type:["string", "integer"]}` is still not allowed in v3 but the former is a valid v3 spec.

**Comment by mbohlool** [comment]
Let me see if I understand the problem correctly. User provides a validation schema that we will use to validate CRs. We want them to be able to provide multiple types. Currently we support OpenAPI v3 schema in CRD validation, so they can do things like '{anyOf: [{type: "string"}, {type: "integer"}]}' but they can't do `{type:["string", "integer"]}`. 

If the above is correct, what if we do a post-processing on the provided spec and convert `{type:["string", "integer"]}` to '{anyOf: [{type: "string"}, {type: "integer"}]}' so it would be a valid OpenAPI v3 spec? I don't like the solution but if we must compromise, it is the closest one to a good solution. I don't like it because it is just a nice to have feature and we are hacking around to let user do it. Also if we don't do it now, it would be always backward compatible to add it.

Regarding `StringArrayOrString`, if the only problem with `IntOrString` is OpenAPI v2 spec, we should be able to proceed with it. I think we are going to support v3 soon (not soon enough for 1.8, but soon). In v2, we should just represent it as `Object`. That means we don't do any validation on it. We are post-processing our spec to do that for some types [here](https://github.com/kubernetes-client/gen/blob/d74ec595ae5fdc7ccd9b35bb997034c25ac1bb77/openapi/preprocess_spec.py#L226).

**Comment by sttts** [comment]
@mbohlool we could do that transformation. But as we restrict ourselves to OpenAPI v3 now in CRD validation, the problem is no problem anymore. I.e. we don't need `StringArrayOrString`. Our approach of validation was that we wanted to accept raw JSON Schema and then refuse certain constructs if they don't fall into OpenAPI v3. But this is essentially cosmetics and only a matter of another error message. So we switched to plain string for `type`, away from `StringArrayOrString`.

**Comment by bgrant0607** [comment]
If we want common tooling, such as get, describe, schema validation, apply, etc., to work across built-in types and CRDs, then CRDs need to follow K8s API conventions, and the provided schema needs to support the same extensions we're adding for the built-in types.

cc @pwittrock

**Comment by sttts** [comment]
> and the provided schema needs to support the same extensions we're adding for the built-in types.

This is on the roadmap, together with a common definition namespace such that you can reference e.g. a PodSpec from a CRD. Would be great to get a proposal done for that early in the 1.9 dev cycle.

**Comment by lavalamp** [comment]
Before looking in detail, let me say that IntOrString is terrible and we should not add more things like it :)

**Comment by lavalamp** [comment]
Why do we need to represent the jsonschema in go types at all? Do we have go code that operates on it? Why not store it in a RawExtension type (i.e., just pass-through the json).

**Comment by lavalamp** [comment]
On IntOrString: when we made it, we thought we were being nice to humans who were going to write  yaml/json by hand directly. But in hindsight, I think it was not worth it. It confuses everyone who touches it. No type spec system represents it well, excepting OpenAPI3, which we don't use yet. Automation systems don't deal with it well, either, e.g. Google's Deployment Manager. It's especially confusing that "12" and 12 weren't the same thing for a while, but now I think they are, but "12%" still isn't.

There was really no good reason not to make a struct {name, number} for port identification. An extra level in a yaml file really isn't a big deal.

Basically, it violates the principle of least surprise for all the automation systems that have to deal with the API, without adding much benefit for users who are directly writing API objects.

**Comment by liggitt** [comment]
this is no longer needed, correct?

**Comment by sttts** [comment]
Correct.
